rust-peer/src/main.rs

```diff
@@ -364,3 +428,91 @@ async fn read_or_create_identity(path: &Path) -> Result<identity::Keypair> {
 
     Ok(identity)
 }
+
+// Simple file exchange protocol
+
+#[derive(Debug, Clone)]
+struct FileExchangeProtocol();
+#[derive(Clone)]
+struct FileExchangeCodec();
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct FileRequest(String);
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileResponse(Vec<u8>);
+
+impl ProtocolName for FileExchangeProtocol {
+    fn protocol_name(&self) -> &[u8] {
+        "/universal-connectivity-file/1".as_bytes()
+    }
+}
+
+#[async_trait]
+impl request_response::Codec for FileExchangeCodec {
+    type Protocol = FileExchangeProtocol;
+    type Request = FileRequest;
+    type Response = FileResponse;
+
+    async fn read_request<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, 1_000_000).await?;
+
+        if vec.is_empty() {
+            return Err(io::ErrorKind::UnexpectedEof.into());
+        }
+
+        Ok(FileRequest(String::from_utf8(vec).unwrap()))
+    }
+
+    async fn read_response<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Response>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, 500_000_000).await?; // update transfer maximum
+
+        if vec.is_empty() {
+            return Err(io::ErrorKind::UnexpectedEof.into());
+        }
+
+        Ok(FileResponse(vec))
```

Comment:
Same here, I think it would be cleaner to just read the entire stream like we do in the cbor protocol:

https://github.com/libp2p/rust-libp2p/blob/73082212374a420d5cb095df6fcf2888249f1dbc/protocols/request-response/src/cbor.rs#L97-L105

Reply:
I've tried to do that, but I got the `BrokenPipe` error when reading the response from the socket. 
```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Kind(BrokenPipe)', src/protocol.rs:69:46
```
https://github.com/youngjoon-lee/universal-connectivity/commit/4e223f03237551764dc2332587fb77b0a5cf54b3

Whereas, JS and Go peers request/receive files with each other successfully. I guess that it's because of the version of rust-libp2p that I use. I'll try it with a newer version of rust-libp2p.

Reply:
@thomaseizinger Do you think that this `BrokenPipe` error is related to the version of rust-libp2p. If so, I'd like to work on this at next PRs, because there are lots of work needed to bump rust-libp2p to the latest, regarding WebRTC. (Currently, this repo uses a [fork](https://github.com/libp2p/universal-connectivity/blob/c83c925d5fb8049c4047c63690e23332ace48332/rust-peer/Cargo.toml#L14) of rust-libp2p and rust-multiaddr. There was an update for WebRTC in rust-multiaddr recently, but I think it's not enough.)